csrc/ir/base_nodes.cpp

```diff
@@ -139,6 +139,12 @@ bool Val::sameAs(const Statement* other) const {
         return false;
       }
     }
+    // non-deterministic operation returns value that wouldn't be the same even
+    // if the inputs are the same.
+    if (!definition_->isDeterministic() ||
```

Comment:
Should `Expr::sameAs` be considered different if one is deterministic and another is not?

Reply:
I think conceptually, that's a good call.

:stuck_out_tongue: For RNGOp, since determinism is embedded as inputs to the op, we didn't need to check that.

I'll add that